crates/holochain/src/core/workflow/incoming_dht_ops_workflow.rs

```diff
@@ -2,109 +2,87 @@
 
 use super::error::WorkflowResult;
 use super::sys_validation_workflow::counterfeit_check;
-use crate::{
-    conductor::{conductor::RwShare, space::Space},
-    core::queue_consumer::TriggerSender,
-};
+use crate::{conductor::space::Space, core::queue_consumer::TriggerSender};
 use holo_hash::DhtOpHash;
 use holochain_sqlite::error::DatabaseResult;
 use holochain_sqlite::prelude::*;
 use holochain_state::prelude::*;
 use holochain_types::dht_op::DhtOp;
 use holochain_types::prelude::*;
+use incoming_ops_batch::InOpBatchEntry;
 use std::{collections::HashSet, sync::Arc};
 use tracing::instrument;
 
-#[cfg(test)]
-mod test;
+mod incoming_ops_batch;
 
-type InOpBatchSnd = tokio::sync::oneshot::Sender<WorkflowResult<()>>;
-type InOpBatchRcv = tokio::sync::oneshot::Receiver<WorkflowResult<()>>;
+pub use incoming_ops_batch::IncomingOpsBatch;
 
-#[derive(Debug)]
-struct InOpBatchEntry {
-    snd: InOpBatchSnd,
-    request_validation_receipt: bool,
-    ops: Vec<(DhtOpHash, DhtOp)>,
+#[cfg(test)]
+mod tests;
+
+struct OpsClaim {
+    incoming_op_hashes: IncomingOpHashes,
+    working_hashes: Vec<DhtOpHash>,
 }
 
-/// A batch of incoming ops memory.
-#[derive(Clone)]
-pub struct IncomingOpsBatch(RwShare<InOpBatch>);
+impl OpsClaim {
```

Comment:
This is new and fixing a potential bug. At the start of the workflow we claim any ops that we're going to work on and prevent any other calls to the workflow processing the same ops at the same time. This is a key protection because we don't hold a database transaction open during the workflow so this is our only protection from duplicate inserts hitting the database.

The problem is that we didn't release those claimed if we exited the workflow early for any reason. This type deals with that by implementing drop so that it's not possible to forget to clean up on workflow exit